### ISSUE TITLE
automatic evaluation for functions of more than one variable

### ISSUE THREAD

**Issue by aklaing** [joaotavora/yasnippet#930]
Hi,

The following snippet allows for the user to input two numbers $4 and $5, and the snippet computes the product $6=$4*$5.  Then after entering a third number $7, the snippet computes a subtraction $8=$6-$7.

```
${1:$$(format-time-string "%Y/%m/%d")} StockSale --- ${2:<comment>}
        ; Stock Ticker:              ${3:<ticker>}
        ; Price                      ${4:0.00}
        ; Number                     ${5:0}
        Equity:StockSale            -${6:0.00$(format "%.2f" (* (string-to-number (yas-field-value 4)) (string-to-number (yas-field-value 5))))}
        Expenses:StockSale:Fee       ${7:0.00}
        Assets:Bank:Checking:Misc    ${8:0.00$(format "%.2f" (- (string-to-number (yas-field-value 6)) (string-to-number (yas-field-value 7))))}

$0

```
It works.

However one thing I don't quite like about it is that you need to tab into fields 6 and 8 in order to get them to be computed/updated to reflect the latest values of $4, $5, and $7 (unlike when you have functions of a single variable implemented as mirrors and they are updated automatically).
Is there a way to make sure fields 6 and 8 are updated immediately whenever any of their dependencies (4,5 and 7) are updated, and also to even avoid having to tab into fields 6 and 8 at all?
I think that would be a much better user experience because

a) Since the user is really unable to modify or override the values of fields 6 and 8, there is no good reason (from a UX perspective) why the user should have to tab into those fields.

b) If the updates to fields 6 and 8 are not computed immediately and automatically whenever one of the dependencies (4, 5 and 7) is modified, then what happens is that you can modify one of the dependencies and then exit the snippet, and in that case
the invariants ($6=$4*$5 and $8=$6-$7) are *not* maintained and that is not good in some situations.

So basically I'd like to know how to evaluate $6 and $8 immediately whenever any character is typed into $4, $5 or $7.  Kindly let me know if there is a way to do this already, or whether it is a worthwhile feature request.

Thank you,

**Comment by npostavs** [comment]
You could do it by turning $6 and $8 into mirrors:

```
${1:$$(format-time-string "%Y/%m/%d")} StockSale --- ${2:<comment>}
        ; Stock Ticker:              ${3:<ticker>}
        ; Price                      ${4:0.00}
        ; Number                     ${5:0}
        Equity:StockSale            -${4:$(format "%.2f" (* (yas-field-as-number 4) (yas-field-as-number 5)))}
        Expenses:StockSale:Fee       ${7:0.00}
        Assets:Bank:Checking:Misc    ${4:$(format "%.2f" (- (* (yas-field-as-number 4) (yas-field-as-number 5)) (yas-field-as-number 7)))}

$0
```

Where `yas-field-as-number` is:

```elisp
(defun yas-field-as-number (field-num)
  (string-to-number (or (yas-field-value field-num) "0")))
```

This does have the downside of repeating the expression `(* (yas-field-as-number 4) (yas-field-as-number 5))` in two places. It could be abstracted into the snippet's environment

```
# -*- mode: snippet -*-
# name: blah
# key: blah
# expand-env: ((equity (lambda () (* (yas-field-as-number 4) (yas-field-as-number 5)))))
# --
${1:$$(format-time-string "%Y/%m/%d")} StockSale --- ${2:<comment>}
        ; Stock Ticker:              ${3:<ticker>}
        ; Price                      ${4:0.00}
        ; Number                     ${5:0}
        Equity:StockSale            -${4:$(format "%.2f" (funcall equity))}
        Expenses:StockSale:Fee       ${7:0.00}
        Assets:Bank:Checking:Misc    ${4:$(format "%.2f" (- (funcall equity) (yas-field-as-number 7)))}

$0
```

Although the `funcall` looks a bit funny.

**Comment by aklaing** [comment]
Thank you for your solution -- it works flawlessly.  TBH I'm quite surprised that it works, because my understanding of the documentation was that if you declare (what was $6) as a mirror of $4, then it will be updated when (only) $4 is updated.  It was not clear to me that it would also be updated when anything else (such as $5) is updated.  Anyway I'm all set, so I'm closing this issue.  Thank you once again!

**Comment by npostavs** [comment]
Yes, practically the only effect of the source number for a mirror transformation is that `yas-text` gets the value of the referenced field. The docs could use some clarification about this.

**Comment by aklaing** [comment]
That would be helpful.  Maybe something along these lines with a very simple example:

Note that all mirror fields are updated when any field is updated.  Moreoever, values of fields other than the field which is mirrored may be obtained with the function yas-field-value.  So for example in the following snippet:
```
firstName=${1:John}
lastName=${2:Smith}
fullName=${1:$(format "%s %s" yas-text (yas-field-value 2))}
```
The full name is updated when either the first name or the last name is updated, in any order.

Thanks

**Comment by aklaing** [comment]
There are two similar google groups that made their lists read-only and changed the description to point to the new location:

[elm-discuss](https://groups.google.com/forum/#!forum/elm-discuss)
[scala-user](https://groups.google.com/forum/#!forum/scala-user)

**If** you do choose to retire the google group, making it read-only may preserve some valuable information that was posted there but not cross-posted on github.

ps.  @joaotavora For some reason I did not receive your reply on the smart-snippet group but thanks!  Yasnippet is a delight to use.

**Comment by npostavs** [comment]
I'm clarifying the docs in #932.
